src/app/qgsapplayertreeviewmenuprovider.cpp

```diff
@@ -316,11 +316,51 @@ QMenu *QgsAppLayerTreeViewMenuProvider::createContextMenu()
         QMenu *menuStyleManager = new QMenu( tr( "Styles" ), menu );
 
         QgisApp *app = QgisApp::instance();
-        menuStyleManager->addAction( tr( "Copy Style" ), app, SLOT( copyStyle() ) );
+        if ( layer->type() == QgsMapLayer::VectorLayer )
+        {
+          QMenu *copyStyleMenu = menuStyleManager->addMenu( tr( "Copy Style" ) );
+          copyStyleMenu->setToolTipsVisible( true );
+          QList<QgsMapLayer::StyleCategory> categories = qgsEnumMap<QgsMapLayer::StyleCategory>().keys();
+          categories.move( categories.indexOf( QgsMapLayer::AllCategories ), 0 ); // move All categories to top
+          for ( QgsMapLayer::StyleCategory category : categories )
+          {
+            QgsMapLayer::ReadableStyleCategory readableCategory = QgsMapLayer::readableStyleCategory( category );
+            QAction *copyAction = new QAction( readableCategory.icon(), readableCategory.name() );
+            copyAction->setToolTip( readableCategory.toolTip() );
+            connect( copyAction, &QAction::triggered, this, [ = ]() {app->copyStyle( layer, category );} );
+            copyStyleMenu->addAction( copyAction );
+            if ( category == QgsMapLayer::AllCategories )
+              copyStyleMenu->addSeparator();
+          }
+        }
+        else
+        {
+          menuStyleManager->addAction( tr( "Copy Style" ), app, SLOT( copyStyle() ) );
```

Comment:
would be better to split this call to avoid the old style connection required by the shortcut addAction method